server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java

```diff
@@ -170,6 +201,37 @@ public Index getWriteIndex() {
         return indices.get(indices.size() - 1);
     }
 
+    public Index getWriteIndex(DocWriteRequest<?> request, Metadata metadata) {
+        if (type != Type.TSDB) {
+            return getWriteIndex();
+        }
+
+        if (request instanceof IndexRequest == false || request.opType() != DocWriteRequest.OpType.CREATE) {
+            return getWriteIndex();
+        }
+
+        // TODO: this really needs be parsed in a streaming manner:
```

Comment:
`XContentParser` now takes a filter. So you can build it with the filter and then do

```
try (XContentParser parser = TS_EXTRACT_CONFIG.parser(request.type(), request.source())) {
  expect(START_OBJECT, parser.next());
  expect(FIELD_NAME, parser.next());
  expect(VALUE, parser.next());
  return DateFieldMaper.DEFAULT_DATE_TIME_FORMATTER.parseMillis(parser.text());
}